quinn/src/lib.rs

```diff
@@ -822,41 +807,50 @@ impl Future for Driver {
                             .pending
                             .entry(connection)
                             .or_insert_with(|| Pending::new(None));
-                        use quinn::Timer::*;
-                        let mut cancel = match timer {
-                            LossDetection => &mut pending.cancel_loss_detect,
-                            Idle => &mut pending.cancel_idle,
-                            Close => unreachable!(),
-                        };
                         let instant = endpoint.epoch + duration_micros(time);
-                        if let Some(cancel) = cancel.take() {
-                            let _ = cancel.send(());
-                        }
-                        let (send, recv) = oneshot::channel();
-                        *cancel = Some(send);
+                        match timer {
+                            Timer::LossDetection => {
+                                if let Some(mut key) = pending.cancel_loss_detect.as_mut() {
+                                    // NOTE: this panics if the key does not exist
+                                    let _ = endpoint.timers.remove(key);
+                                    *key = endpoint
+                                        .timers
+                                        .insert_at((connection, Timer::LossDetection), instant);
```

Comment:
It could, but this logic was wrong. With the fixed version it may not make as much sense:

```rust
                                if let Some(mut key) = pending.cancel_loss_detect.as_ref() {
                                    endpoint.timers.remove(key);
                                }
                                let key = endpoint
                                    .timers
                                    .insert_at((connection, Timer::LossDetection), instant);
                                pending.cancel_loss_detect = Some(key)
```

vs 

```rust
                                if let Some(mut key) = pending.cancel_loss_detect.as_ref() {
                                    endpoint.timers.reset_at(key, instant);
                                } else {
                                    let key = endpoint
                                        .timers
                                        .insert_at((connection, Timer::LossDetection), instant);
                                    pending.cancel_loss_detect = Some(key)
                                }
```

Reply:
I'm surprised to discover that there's no way to copy a `delay_queue::Key`. I'll work with upstream to get a `Clone` impl added at minimum, but in the meantime this is fine.